python/ray/tune/trial_executor.py

```diff
@@ -0,0 +1,146 @@
+        in error, but no exception will be thrown.
+
+        Args:
+            error (bool): Whether to mark this trial as terminated in error.
+            error_msg (str): Optional error message.
+            stop_logger (bool): Whether to shut down the trial logger.
+        """
+        raise NotImplementedError('subclasses of TrialExecutor must provide '
+                                  'stop_trial() method')
+
+    def restart_trial(self, trial, error_msg=None):
+        """Restarts the trial.
+
+        The state of the trial should restore from the last checkpoint.
+
+        Args:
+            error_msg (str): Optional error message.
+        """
+        try:
+            print("Attempting to recover trial state from last checkpoint")
+            self.stop_trial(trial, error=True, error_msg=error_msg,
+                            stop_logger=False)
+            trial.result_logger.flush()
+            self.start_trial(trial)
+        except Exception:
+            error_msg = traceback.format_exc()
+            print("Error recovering trial from checkpoint, abort:", error_msg)
+            self.stop_trial(trial, error=True, error_msg=error_msg)
+
+    def on_scheduler_decision(self, trial, decision):
+        """A hook called when TrialScheduler make a decision.
+
+        Args:
+            decision (str): (CONTINUE,PAUSE,STOP) constant from TrialScheduler.
+        """
+        pass
+
+    def pause_trial(self, trial):
+        """Pauses the trial.
+
+        We want to release resources (specifically GPUs) when pausing an
+        experiment. This results in PAUSED state that similar to TERMINATED.
+        """
+        assert trial.status == Trial.RUNNING, trial.status
+        try:
+            self.save(trial, Checkpoint.TYPE_OBJECT_STORE)
+            self.stop_trial(trial, stop_logger=False)
+            trial.status = Trial.PAUSED
+        except Exception:
+            print("Error pausing runner:", traceback.format_exc())
+            trial.status = Trial.ERROR
+
+    def unpause_trial(self, trial):
+        """Sets PAUSED trial to pending to allow scheduler to start."""
+        assert trial.status == Trial.PAUSED, trial.status
+        trial.status = Trial.PENDING
+
+    def resume_trial(self, trial):
+        """Resumes PAUSED trials. This is a blocking call."""
+        assert trial.status == Trial.PAUSED, trial.status
+        self.start_trial(trial)
+
+    def get_running_trials(self):
+        """Returns all running trials."""
+        raise NotImplementedError('subclasses of TrialExecutor must provide '
+                                  'get_running_trials() method')
+
+    def on_step_begin(self):
+        """A hook called before running one step of the trial event loop."""
+        pass
+
+    def on_step_end(self):
+        """A hook called after running one step of the trial event loop."""
+        pass
+
+    def fetch_one_result(self):
```

Comment:
Maybe we could add a TODO of `fetch_mutliple_results` to allow implementations of efficient polling mechanism